internal/vulnstore/postgres/updatevulnerabilities.go

```diff
@@ -0,0 +1,192 @@
+		dist_pretty_name, repo_name, repo_key, repo_uri, fixed_in_version, version_kind, vulnerable_range
+	) VALUES (
+	  $1, $2, $3, $4, $5, $6, $7, $8, $9, $10,
+	  $11, $12, $13, $14, $15, $16, $17, $18,
+	  $19, $20, $21, $22, $23, $24, VersionRange($25, $26)
+	)
+	ON CONFLICT (hash_kind, hash) DO NOTHING
+	RETURNING id)
+INSERT INTO uo_vuln (uo, vuln) VALUES (
+	$27,
+	COALESCE (
+		(SELECT id FROM attempt),
+		(SELECT id FROM vuln WHERE hash_kind = $1 AND hash = $2)));`
+	)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "internal/vulnstore/postgres/updateVulnerabilities").
+		Logger()
+	ctx = log.WithContext(ctx)
+	tx, err := pool.Begin(ctx)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("unable to start transaction: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	var id uint64
+	var ref uuid.UUID
+	if err := pool.QueryRow(ctx, create, updater, string(fingerprint)).Scan(&id, &ref); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to create update_operation: %w", err)
+	}
+	log.Debug().
+		Str("ref", ref.String()).
+		Msg("update_operation created")
+
+	// batch insert vulnerabilities
+	skipCt := 0
+	mBatcher := microbatch.NewInsert(tx, 2000, time.Minute)
+	for _, vuln := range vulns {
+		if vuln.Package == nil || vuln.Package.Name == "" {
+			skipCt++
+			continue
+		}
+
+		pkg := vuln.Package
+		dist := vuln.Dist
+		repo := vuln.Repo
+		if dist == nil {
+			dist = &zeroDist
+		}
+		if repo == nil {
+			repo = &zeroRepo
+		}
+		hashKind, hash := md5Vuln(vuln)
+		vKind, vrLower, vrUpper := rangefmt(vuln.Range)
+
+		err := mBatcher.Queue(ctx, insert,
+			hashKind, hash,
+			vuln.Name, vuln.Updater, vuln.Description, vuln.Links, vuln.Severity, vuln.NormalizedSeverity,
+			pkg.Name, pkg.Version, pkg.Kind,
+			dist.DID, dist.Name, dist.Version, dist.VersionCodeName, dist.VersionID, dist.Arch, dist.CPE, dist.PrettyName,
+			repo.Name, repo.Key, repo.URI,
+			vuln.FixedInVersion, vKind, vrLower, vrUpper,
+			id,
+		)
+		if err != nil {
+			return uuid.Nil, fmt.Errorf("failed to queue vulnerability: %w", err)
+		}
+	}
+	if err := mBatcher.Done(ctx); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to finish batch vulnerability insert: %w", err)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to commit transaction: %w", err)
+	}
+	log.Debug().
+		Str("ref", ref.String()).
+		Int("skipped", skipCt).
+		Int("inserted", len(vulns)-skipCt).
+		Msg("update_operation committed")
+	return ref, nil
+}
+
+// Md5Vuln creates an md5 hash from the members of the passed-in Vulnerability,
+// giving us a stable, context-free identifier for this revision of the
+// Vulnerability.
+func md5Vuln(v *claircore.Vulnerability) (string, []byte) {
+	h := md5.New()
+	h.Write([]byte(v.Name))
+	h.Write([]byte(v.Description))
+	h.Write([]byte(v.Links))
+	h.Write([]byte(v.Severity))
+	if v.Package != nil {
+		h.Write([]byte(v.Package.Name))
```

Comment:
this performs a copy for every field value, including descriptions. 
it's also in the hot path of a write. I think we should use go's optimization: http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#string_byte_slice_conv

Reply:
I think I can solve this by just using a `bytes.Buffer`. Looking at the code, it's just a `copy` in the happy path.

Reply:
Cool, a single copy is better then copying each field. If that strikes a balance between readability and perf, would be a good move.